File: LeetCode/Medium/0116_populating-next-right-pointers-in-each-node/20240627-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

from collections import deque


class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root is None:
            return None

        queue = deque([root])
        count = 0
        count_limit = 1
        prev_node = None

        while len(queue) != 0:
            node = queue.popleft()
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)

            count += 1
            if prev_node is not None:
                prev_node.next = node
            prev_node = node

            if count == count_limit:
                count = 0
                count_limit *= 2
                prev_node = None

        return root

File: LeetCode/Medium/0106_construct-binary-tree-from-inorder-and-postorder-traversal/20240627-1.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

impl Solution {
    pub fn build_tree(inorder: Vec<i32>, mut postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let inorder_index_map = create_index_map(&inorder);
        build(
            0,
            inorder.len() as i32 - 1,
            &mut postorder,
            &inorder_index_map,
        )
    }
}

fn build(
    inorder_l: i32,
    inorder_r: i32,
    postorder: &mut Vec<i32>,
    inorder_index_map: &HashMap<i32, i32>,
) -> Option<Rc<RefCell<TreeNode>>> {
    if inorder_l > inorder_r {
        return None;
    }

    let mut root = TreeNode::new(postorder.pop().unwrap());
    let index = *inorder_index_map.get(&root.val).unwrap();

    root.right = build(index + 1, inorder_r, postorder, inorder_index_map);
    root.left = build(inorder_l, index - 1, postorder, inorder_index_map);

    Some(Rc::new(RefCell::new(root)))
}

fn create_index_map(arr: &[i32]) -> HashMap<i32, i32> {
    let mut result = HashMap::new();

    for i in 0..arr.len() {
        result.insert(arr[i], i as i32);
    }

    result
}

struct Solution;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
